### PR TITLE
Added auto hinting compensation to fix too wide gaps between small glyphs

### PR DESCRIPTION
**Pull request by oomek**
* [x] Has this change been discussed on [the forum](https://en.sfml-dev.org/forums/index.php#c3) or in an issue before?
* [x] Does the code follow the SFML [Code Style Guide](https://www.sfml-dev.org/style.php)?
* [ ] Have you provided some example/test code for your changes?
* [ ] If you have additional steps which need to be performed list them as tasks!

----

## Description

Forced Autohint requires left and right bearings to be taken into acccount during kerning calculation. Without this the gaps between small glyph sizes are twice as wide on a lot of fonts.

This PR is related to the issue https://github.com/SFML/SFML/issues/1642#issuecomment-763700187

## Tasks

* [ ] Tested on Linux
* [x] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

### CODE REVIEW

File: src/SFML/Graphics/Font.cpp [hunk 1]
```diff
@@ -375,30 +376,37 @@ bool Font::hasGlyph(Uint32 codePoint) const
...
 
+        // Retrieve position compensation deltas generated by FT_LOAD_FORCE_AUTOHINT flag
+        float firstRsbDelta = getGlyph(first, characterSize, bold).rsb_delta;
+        float secondLsbDelta = getGlyph(second, characterSize, bold).lsb_delta;
+
         // Get the kerning vector
         FT_Vector kerning;
-        FT_Get_Kerning(face, index1, index2, FT_KERNING_DEFAULT, &kerning);
+        kerning.x = kerning.y = 0;
+        if( FT_HAS_KERNING(face))
```
Comment (eXpl0it3r) [inline src/SFML/Graphics/Font.cpp:400]: Should be `if (`

Comment (oomek) [reply src/SFML/Graphics/Font.cpp]: @eXpl0it3r Thanks for spotting that. I'm usually pedantic when it comes to formatting conventions.

File: include/SFML/Graphics/Glyph.hpp [hunk 2]
```diff
@@ -52,6 +52,8 @@ class SFML_GRAPHICS_API Glyph
     // Member data
     ////////////////////////////////////////////////////////////
     float     advance;     //!< Offset to move horizontally to the next character
+    int       lsb_delta;   //!< Left offset after forced autohint. Internally used by getKerning()
+    int       rsb_delta;   //!< Right offset after forced autohint. Internally used by getKerning()
```
Comment (LaurentGomila) [inline include/SFML/Graphics/Glyph.hpp:56]: These two members don't follow our naming convention (should be `lsbDelta` and `rsbDelta`).

File: src/SFML/Graphics/Font.cpp [hunk 3]
```diff
@@ -375,30 +376,37 @@ bool Font::hasGlyph(Uint32 codePoint) const
...
+float Font::getKerning(Uint32 first, Uint32 second, unsigned int characterSize, bool bold) const
 {
     // Special case where first or second is 0 (null character)
     if (first == 0 || second == 0)
         return 0.f;
 
     FT_Face face = static_cast<FT_Face>(m_face);
 
-    if (face && FT_HAS_KERNING(face) && setCurrentSize(characterSize))
+    if (face && setCurrentSize(characterSize))
```
Comment (LaurentGomila) [inline src/SFML/Graphics/Font.cpp:387]: Since this block now handles the "no kerning" case as well, the comment in the "else" block should be adjusted.

**Review by LaurentGomila — COMMENTED** [review]
Looks ok, except some minor adjustments

File: src/SFML/Graphics/Font.cpp (hunk 3 above)
Comment (oomek) [reply src/SFML/Graphics/Font.cpp:387]: Shall I just leave `invalid font` or provide something more meaningful?

**Comment by oomek** [comment]
Regarding that added function parameter:
https://github.com/SFML/SFML/blob/d2406fcff21947e4ad125f40f1234787b40b2cee/include/SFML/Graphics/Font.hpp#L220
Shall I update the brief as well? If so, do we need to indicate that when different bold flag is used than already cached the glyph will be loaded?

**Comment by eXpl0it3r** [comment]
> Shall I update the brief as well?

If you think the brief could need an update/change, then certainly.
What you certainly need to add is `/param` for the new parameter.

> If so, do we need to indicate that when different bold flag is used than already cached the glyph will be loaded?

Does the user of the function need to care about this implementation detail?
If only in some very special cases, then no, I wouldn't add that information.

**Comment by oomek** [comment]
Maybe it's just me being paranoid, but I found 2 fonts out of 50 I tested with unusually narrow whitespace.
It's not relevant to kerning compensation, but how `whitespaceWidth` is calculated. Maybe it's also a good opportunity to address it aswell by clamping the whitespaceWidth to at least 2 pixels by:
```
    float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance;
    float letterSpacing   = ( whitespaceWidth / 3.f ) * ( m_letterSpacingFactor - 1.f );
    whitespaceWidth      += letterSpacing;
    if (whitespaceWidth < 2.f) whitespaceWidth = 2.f;
```
Here is the screenshot of the affected font before and after:
image
Please open it in 100% browser zoom to avoid any potential scaling.

File: src/SFML/Graphics/Font.cpp (hunk 3 above)
Comment (LaurentGomila) [reply src/SFML/Graphics/Font.cpp:387]: Whatever. Even removing the comment would be fine.

**Comment by oomek** [comment]
Would you like me to commit that space width compensation as well?

**Comment by LaurentGomila** [comment]
Since it's a different issue, let's address it in another PR.

**Comment by oomek** [comment]
Half of the CI tasks did not start. Do they have to be triggered manually?

**Comment by eXpl0it3r** [comment]
SFMLBuildMaster: Build this please.

**Comment by eXpl0it3r** [comment]
> Half of the CI tasks did not start. Do they have to be triggered manually?
Our own CI setup only executes on demand on third-party PRs for security reasons.
I have since triggered them.

@LaurentGomila any additional comment for the PR?

**Review by LaurentGomila — APPROVED** [review]

**Pull request merged**